src/cache.jl

```diff
@@ -34,22 +87,20 @@ const specialization_counter = Ref{UInt}(0)
     resize!(new_ci.linetable, 1)                # see note below
     empty!(new_ci.ssaflags)
     new_ci.ssavaluetypes = 0
+    new_ci.min_world = min_world[]
+    new_ci.max_world = max_world[]
     new_ci.edges = MethodInstance[mi]
     # XXX: setting this edge does not give us proper method invalidation, see
     #      JuliaLang/julia#34962 which demonstrates we also need to "call" the kernel.
     #      invoking `code_llvm` also does the necessary codegen, as does calling the
     #      underlying C methods -- which GPUCompiler does, so everything Just Works.
 
     # prepare the slots
-    new_ci.slotnames = Symbol[Symbol("#self#"), :cache, :job, :compiler, :linker]
-    new_ci.slotflags = UInt8[0x00 for i = 1:5]
-    cache = SlotNumber(2)
-    job = SlotNumber(3)
-    compiler = SlotNumber(4)
-    linker = SlotNumber(5)
-
-    # call the compiler
-    push!(new_ci.code, ReturnNode(id))
+    new_ci.slotnames = Symbol[Symbol("#self#"), :ft, :tt]
+    new_ci.slotflags = UInt8[0x00 for i = 1:3]
+
+    # return the world
+    push!(new_ci.code, ReturnNode(world))
```

Comment:
Would it be fine to return the generator's world (post-https://github.com/JuliaLang/julia/pull/48766)? That's what I'm doing in https://github.com/JuliaGPU/GPUCompiler.jl/pull/403.

Returning the `mi` could work I guess, but would need some refactoring. We're currently returning the world so that we can use it to construct a FunctionSpec, which is the object that we use to look-up the code to compile: https://github.com/JuliaGPU/GPUCompiler.jl/blob/7e0a6ff485f17886a31bbae158a63bcc13d41933/src/interface.jl#L62-L71

Reply:
Note that the reason we use a separate struct is that it used to carry more fields. Now that they're gone, maybe we should just use MethodInstance instead of FunctionSpec, as @vchuravy notes.

Reply:
No, it is never allowable to expose the user to worlds from the compiler and vice versa. It violates the contract that only edges are observable effects.

Reply:
OK. We do however need the world age for some cases though, e.g., to look-up in our CodeInstance cache (should I use `WorldView(mi.def.primary_world, mi.def.deleted_world)` here?), and to pass to `jl_create_native` (not sure what to use here). What should I use instead of the world that we now return here?